### PR TITLE
[RFC] Implement caching for user defined triton kernels

### PR DESCRIPTION
**Pull request by oulgen**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #140326

This PR adds caching for user defined triton kernels by putting the transitive closure of source code in node.meta along with constant arguments.

One HUGE hack we do here is a node looks like
```
triton_kernel_wrapper_functional_proxy = torch.ops.higher_order.triton_kernel_wrapper_functional(kernel_idx = 0, constant_args_idx = 1, grid = [(1, 1, 1)], tma_descriptor_
metadata = {}, kwargs = {'in_ptr0': arg0_1, 'in_ptr1': arg1_1, 'out_ptr': arg0_1}, tensors_to_clone = ['out_ptr']);
```
so we use regex to remove `kernel_idx = 0, constant_args_idx = 1` parts as they are not relevant to cache hash. This is horrible and I'd like to eventually not use pickle as a hashing alternative but this is a longer project.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

Differential Revision: [D65895744](https://our.internmc.facebook.com/intern/diff/D65895744)

### CODE REVIEW

File: torch/_inductor/codegen/wrapper.py
Comment (ezyang) [inline torch/_inductor/codegen/wrapper.py:1556]: any chance of doing this code move as a refactor pr stacked before this? Hard to review like this

File: torch/_higher_order_ops/triton_kernel_wrap.py
Comment (zou3519) [inline torch/_higher_order_ops/triton_kernel_wrap.py:766]: We should change how the constant args works (to also benefit the triton kernel export work). Constant args should be inputs to the node, unless they are non-fx-graphable types. The non-fx-graphable types should go into the side table.

File: torch/_inductor/codecache.py
Comment (zou3519) [inline torch/_inductor/codecache.py:637]: maybe these should just both be hashes so we don't need to do the regex delete thing

Comment (zou3519) [reply torch/_inductor/codecache.py:637]: how does fx_graph_cache handle graphs that look similar but have different variable names? I think it's reasonable to have some sort of "graph normalization pass"

File: torch/_inductor/codecache.py
Comment (zou3519) [inline torch/_inductor/codecache.py:821]: @oulgen We've got some options:
- toss the triton kernel source into node.meta
- toss the triton kernel source into graph.meta
- toss the triton kernel source as a string input to the node (we don't need to read from the source if the triton kernel is still alive; we can also probably develop some mechanism for this to print in a sane way).

Ditto for the constant args (in option 3, if the constant args are not fx-graphable, then we could just say they're not cacheable).

One question is: how much do you care about the triton kernel export serializability work? There could be one solution that works for both triton kernel export serializability and the caching work.

Comment (zou3519) [reply torch/_inductor/codecache.py:821]: I kind of like storing the triton kernel source in the graph.meta -- the claim here is that we are essentially going to move the side table to be in the graph itself.

File: torch/_higher_order_ops/triton_kernel_wrap.py
Comment (oulgen) [reply torch/_higher_order_ops/triton_kernel_wrap.py:766]: I guess I didnt want constant args to be spit between on graph and on side table, wanted to keep them all together originally

File: torch/_inductor/codecache.py
Comment (oulgen) [reply torch/_inductor/codecache.py:637]: it does not, unless the graph are identical, we cache miss. I thought this was the safest starting point. We could do some normalization eventually but i want to understand how much does this bring us

File: torch/_inductor/codecache.py
Comment (oulgen) [reply torch/_inductor/codecache.py:821]: I think what you're suggesting is a big refactor to get rid of side table all together which likely requires us to eval the source code for inductor so that we have a function pointer again, or store all the kernel properties ahead of time. For caching purposes only, i wanted to start with just the source since we dont need to restore anything but only use it as a key.

The problem with storing in graph.meta is if someone dropped the meta or removed one in an opt pass, we can no longer assert that we have all of them, as opposed to the current assertion. wdyt?

File: torch/_inductor/codegen/wrapper.py
Comment (oulgen) [reply torch/_inductor/codegen/wrapper.py]: done

**Comment by oulgen** [comment]
@oulgen has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D65895744).

File: torch/_inductor/codecache.py
Comment (zou3519) [reply torch/_inductor/codecache.py:820]: My suggestion is more: I'm worried that we're going to end up with two different implementations, one for caching and one for serialization, or that when we implement serialization we will end up re-implementing caching. Since the timeline for serialization is also soon (@ydwu4's planning on working on it in the next couple of weeks?) I wonder if it makes sense to slow down and figure out how to co-design these. Feel free to make an argument that the code in this PR is easy to rip out, for example.

I'm not suggesting that we should eval the source code for inductor -- I can see how that would be undesirable.

> The problem with storing in graph.meta is if someone dropped the meta or removed one in an opt pass, we can no longer assert that we have all of them, as opposed to the current assertion. wdyt?

In either (1) or (2) it seems like we need to build some assertions into the system. For node.meta we would need to build assertions that for all triton_kernel_* nodes, the node.meta must be populated. For graph.meta, we would need to build the assertion that if the graph has any triton_kernel_* nodes, the triton kernels they are referring to have their sources in graph.meta. A benefit of building into graph.meta is that we can deduplicate nodes (idk if that actually matters for things like cache key computation)

File: torch/_higher_order_ops/triton_kernel_wrap.py
Comment (zou3519) [reply torch/_higher_order_ops/triton_kernel_wrap.py:789]: We already split the constant args to do triton kernel non-strict export, because non-strict export doesn't have the concept of "constant args"

File: torch/_inductor/codecache.py
Comment (aakhundov) [reply torch/_inductor/codecache.py:820]: @zou3519 I agree eval is undesirable (e.g., as we'd need to collect all transitive dependencies, relevant imports etc.; I think, we rely on `Autotuner` and `JITFunction` instances in multiple places, so we'd need to be able to reliably recreate them from the source).

> I'm not suggesting that we should eval the source code for inductor -- I can see how that would be undesirable.

Given just the source, how can we then reconstruct the kernel instance w/o eval?

Comment (aakhundov) [reply torch/_inductor/codecache.py:820]: Did we consider serializing the side table instead of the kernel source? Serialized Triton kernel source (in Python) makes little sense w/o Python runtime, and with it we could also deserialize the kernel instances in the side table?

Comment (oulgen) [reply torch/_inductor/codecache.py:820]: @zou3519 yep your points make sense, but I do worry that this requires a larger refactor around user defined triton kernels, and i'm primarily trying to unblock ads models as their compile time reduction is blocked on lack of user defined triton kernel caching. Would you be willing to proceed with this and later rip this out or move the assertions to graph? Moving the assertions to graph instead of node is trickier which is why i chose this for the quick-and-dirty solution. If you think this is too much tech debt, I'm happy to accept that.

Comment (oulgen) [reply torch/_inductor/codecache.py:820]: @aakhundov Deserialization is the hard question here, esp if we want to do that without eval, by storing a ton of metadata, and since I didnt need it for hashing, I tried to avoid it. But I dont think serde of the side table is any harder than serde of a single kernel pointer, so i think it boils down to the same problem.

Comment (zou3519) [reply torch/_inductor/codecache.py:820]: >  i'm primarily trying to unblock ads models as their compile time reduction is blocked on lack of user defined triton kernel caching. Would you be willing to proceed with this and later rip this out or move the assertions to graph?

Yes I'm happy to proceed with the current design, esp bc there is a pressing need for this right now

**Comment by oulgen** [comment]
@oulgen has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D65895744).

File: torch/_inductor/codecache.py
Comment (oulgen) [reply torch/_inductor/codecache.py:637]: idx is just an entry into the table, we graph that entry and convert it to source code. we add the transitive closure of source code of each triton kernel in this graph, we assert that we do it for each triton op on the graph.

Comment (ezyang) [reply torch/_inductor/codecache.py:637]: Err, isn't the real reason because you shoved it in `user_defined_triton_kernel_source_and_constant_args`

Comment (oulgen) [reply torch/_inductor/codecache.py:637]: Yea, that's what I said, no? grab the entry put it into node.meta and later put it in the cache key

Comment (zou3519) [reply torch/_inductor/codecache.py:637]: I think you guys are saying the same thing lol

File: torch/_functorch/_aot_autograd/autograd_cache.py
Comment (zou3519) [inline torch/_functorch/_aot_autograd/autograd_cache.py:278]: Are you enabling this for AOTAutogradCache too, or is that another PR? 

If so, Dynamo produces triton_kernel_wrapper_mutation or a triton_op and I don't see those being handling here

File: test/inductor/test_codecache.py
Comment (zou3519) [inline test/inductor/test_codecache.py:527]: For testing:
- can we test that calling a function with a different triton kernel results in a cache miss?

File: torch/_higher_order_ops/triton_kernel_wrap.py
Comment (zou3519) [inline torch/_higher_order_ops/triton_kernel_wrap.py:641]: I'm confused, FXGraphCache should only see one of {triton_kernel_wrapper_mutation, triton_kernel_wrapper_functional} (depending on where the cache is in Inductor), did we need to make both cacheable?

**Review by zou3519 — COMMENTED** [review]
Given our discussion, the approach LGTM. I want to check if we're changing the behavior for AOTAutogradCache since there are some code changes there, also added some testing suggestions

**Comment by ezyang** [comment]
Deferring to Richard for approval

File: torch/_functorch/_aot_autograd/autograd_cache.py
Comment (oulgen) [reply torch/_functorch/_aot_autograd/autograd_cache.py:278]: This PR is just FXGraphCache, I needed to update the pickling so this is just conforming the signature. I can do AOTAutograd as a follow up

**Comment by oulgen** [comment]
@zou3519 added unit test for cache miss, will do aotautograd as a follow up

File: torch/_higher_order_ops/triton_kernel_wrap.py
Comment (zou3519) [reply torch/_higher_order_ops/triton_kernel_wrap.py:641]: No harm, the problem is mostly that the test coverage only covers one of them so if we didn't actually implement the other one correctly and it somehow shows up in in the future then problems

**Review by zou3519 — APPROVED** [review]

**Comment by oulgen** [comment]
@oulgen has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D65895744).

**Comment by oulgen** [comment]
@oulgen has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D65895744).

**Pull request closed without merging**
